File: 2021/3/3.py
from copy import deepcopy
import re

#Test data
with open('test.txt', 'r') as f:
    test_bit_strings = [f.strip() for f in f.readlines()]

#Actual data
with open('input.txt', 'r') as f:
    bit_strings = [f.strip() for f in f.readlines()]

def most_common(lst):
    return max(lst, key=lst.count)

def gamma_rate(data):
    '''Gets the most common bit in each string index for any number
    bit strings of any length'''
    bit_map = {indx: [] for indx, b in enumerate(data[0])}
    for bit_string in data:
        for indx, bit in enumerate(bit_string):
            bit_map[indx].append(bit)
    gamma_string = ''
    for i in bit_map.values():
        gamma_string += most_common(i)
    gamma_number = int(gamma_string, base=2)
    return gamma_string, gamma_number

def epsilon_rate(data):
    '''Opposite of gamma rate, least common bit at each index for a
    number of bit strings'''
    mirror = gamma_rate(data)[0]
    epsilon_string = ''
    for i in mirror:
        if i == '1':
            epsilon_string += '0'
        elif i == '0':
            epsilon_string += '1'
    epsilon_number = int(epsilon_string, base=2)
    return epsilon_string, epsilon_number

gam = gamma_rate(test_bit_strings)
ep = epsilon_rate(test_bit_strings)
print(gam, ep, gam[1] * ep[1])

def oxygen_rating(data):
    most_common_bits = gamma_rate(data)[0]
    considering = deepcopy(data)
    for b in most_common_bits:
        r = re.compile('^' + most_common_bits[:most_common_bits.index(b)])
        considering = list(filter(r.match, considering))
    return considering

print(oxygen_rating(test_bit_strings))

File: 2021/3/3.js
const fs = require('fs');

//Test data
const testBitStrings = fs.readFileSync('test.txt', 'utf8').split('\n').map(str => str.trim()).filter(str => str !== '');

//Actual data
const bitStrings = fs.readFileSync('input.txt', 'utf8').split('\n').map(str => str.trim()).filter(str => str !== '');

function mostCommon(lst) {
  return lst.reduce((a, b, i, arr) => (arr.filter(v => v===a).length >= arr.filter(v => v===b).length ? a : b), null);
}

function gammaRate(data) {
  let bitMap = data[0].split('').map(() => []);
  for (const bitString of data) {
    for (const [indx, bit] of bitString.split('').entries()) {
      bitMap[indx].push(bit);
    }
  }
  const gammaString = bitMap.map(mostCommon).join('');
  const gammaNumber = parseInt(gammaString, 2);
  return [gammaString, gammaNumber];
}

function epsilonRate(data) {
  const mirror = gammaRate(data)[0].split('').map(bit => bit === '1' ? '0' : '1').join('');
  const epsilonNumber = parseInt(mirror, 2);
  return [mirror, epsilonNumber];
}

function oxygenRating(data) {
  let oxygenData = [...data];
  let count = 0;
  while (oxygenData.length > 1) {
    let zeros = 0, ones = 0;
    for (const bitString of oxygenData) {
      if (bitString[count] === '0') {
        zeros += 1;
      } else if (bitString[count] === '1') {
        ones += 1;
      }
    }
    if (zeros > ones) {
      oxygenData = oxygenData.filter(bitString => bitString[count] === '0');
    } else {
      oxygenData = oxygenData.filter(bitString => bitString[count] === '1');
    }
    count += 1;
  }
  const oxygenNumber = parseInt(oxygenData[0], 2);
  return [oxygenData, oxygenNumber];
}

function co2Rating(data) {
  let co2Data = [...data];
  let count = 0;
  while (co2Data.length > 1) {
    let zeros = 0, ones = 0;
    for (const bitString of co2Data) {
      if (bitString[count] === '0') {
        zeros += 1;
      } else if (bitString[count] === '1') {
        ones += 1;
      }
    }
    if (zeros > ones) {
      co2Data = co2Data.filter(bitString => bitString[count] === '1');
    } else {
      co2Data = co2Data.filter(bitString => bitString[count] === '0');
    }
    count += 1;
  }
  const co2Number = parseInt(co2Data[0], 2);
  return [co2Data, co2Number];
}

const gam = gammaRate(testBitStrings);
const ep = epsilonRate(testBitStrings);
console.log(gam, ep, gam[1] * ep[1]);

const o = oxygenRating(bitStrings)[1];
const c = co2Rating(bitStrings)[1];
console.log(o * c);